uwblueprint/planet-read
# Allow users to only translate and review one story at a time per language

## Notion ticket link



[Allow users to only translate and review one story at a time per language](https://www.notion.so/uwblueprintexecs/Allow-users-to-only-translate-and-review-one-story-at-a-time-per-language-b0b40542f8f44f47bf3490e03004491e)



## Implementation description
_backend changes_:
- Added checks before assigning user as a translator/reviewer to ensure they have not already been assigned to translate/review a story of the same language
- Modified the `get_story_translations` query parameters to all be optional
- Updated `insert_test_data.py` script
_frontend changes_:
- Updated frontend to query for and display all stories user is currently translating/reviewing in homepage "My Work" + disabled filters 
- Added translator/reviewer pill to story card



## Steps to test

1. Rebuild docker containers; clear +re-seed database:
```
$ docker exec -it planet-read_py-backend_1 /bin/bash -c "python -m tools.insert_test_data erase"
$ docker exec -it planet-read_py-backend_1 /bin/bash -c "python -m tools.insert_test_data"
```
2. Login to frontend (these steps will use Dwight D. Eisenhower as the example user)
3. Verify that in the "My Work" tab, all the stories the user is translating/reviewing show up:
  - Pride and Prejudice (English UK) [Translator]
  - To Kill a Mockingbird (German) [Translator]
  - The Great Gatsby (English US) [Reviewer]
![image](https://user-images.githubusercontent.com/32009013/134185066-4ea0258d-b7c2-41a4-a72f-223bf8eaf827.png)
4. Check that you are able to click into any of the stories and view the translation/review platform without issues
5. Click on browse stories and try to translate a story the user has level access for, but in a language the user is already translating; this should result in an error `Error: User can't be assigned as a translator`
  - For Eisenhower try English UK <= level 4 or German <= level 3
![image](https://user-images.githubusercontent.com/32009013/134189002-7830dca2-07ae-4ec7-a76d-a27f4497976f.png)
6. Try to translate a story the user has level access for, but in a language the user is *not* currently translating
  - Try a story in English US <= level 4
7. Repeat steps 5 and 6 for reviewing stories
  - Eisenhower can review English UK <= level 3 (this should work without issues) and English US <= level 4 (this should fail as he is already reviewing a story in this language)
8. Navigate back to the "My Work" tab and verify that any stories the user was successfully added as a translator/reviewer appear with the correct role pill 


![disabled filter](https://user-images.githubusercontent.com/32009013/134185113-86fbda01-3634-47e3-80e9-fdf85f08ef14.gif)


## What should reviewers focus on?

- Design of the story card role pill/disabled filters
- Checks added to `story_service.py`
- Data seeded in `insert_test_data.py` is consistent (i.e. users seeded in db are only translating and reviewing at most 1 story per language)
- Is `seed_db.sh` still being used by the team? Should I update the data there as well?

## Checklist
